sockeye/rnn_attention.py

```diff
@@ -343,33 +364,65 @@ def attend(att_input: AttentionInput, att_state: AttentionState) -> AttentionSta
             attention_scores = mx.sym.batch_dot(lhs=source_hidden, rhs=expanded_decoder_state,
                                                 name="%sbatch_dot" % self.prefix)
 
-            context, attention_probs = get_context_and_attention_probs(source, source_length, attention_scores)
+            context, attention_probs = get_context_and_attention_probs(source, source_length, attention_scores, self.dtype)
             return AttentionState(context=context,
                                   probs=attention_probs,
                                   dynamic_source=att_state.dynamic_source)
 
         return attend
 
 
+@Attention.register(C.ATT_DOT_SCALED)
+class DotScaledAttention(DotAttention):
+    """
+    DotAttention with optional scale before dot product [Vaswani et al, 2017].
+
+    :param input_previous_word: Feed the previous target embedding into the attention mechanism.
+    :param source_num_hidden: Number of hidden units in source.
+    :param query_num_hidden: Number of hidden units in query.
+    :param num_hidden: Number of hidden units.
+    :param scale: Optionally scale query before dot product [Vaswani et al, 2017].
+    :param dtype: data type.
+    """
+
+    def __init__(self,
+                 input_previous_word: bool,
+                 source_num_hidden: int,
+                 query_num_hidden: int,
+                 num_hidden: int,
+                 scale: Optional[float] = None,
+                 dtype: Optional[str] = C.DTYPE_FP32) -> None:
+        super().__init__(input_previous_word,
+                         source_num_hidden,
+                         query_num_hidden,
+                         num_hidden,
+                         dtype=dtype)
+        self.scale = scale if scale else num_hidden ** -0.5
```

Comment:
if you kept scale as a parameter on `DotAttention` you could pass it directly in the constructor.

Reply:
As discussed, `scale` parameter will be now supported in `DotScaleAttantion`, but not `DotAttention`

Reply:
I agree with Tobi's comment. Its confusing to set self.scale to None in `DotAttention` just to have inheritance. I would be in favor of a single `DotAttention` class that takes a scale parameter (that can be None).

Reply:
I would leave it as is: scaled dot attention is still conceptually dot attention, so it makes sense to have a single class DotAttention that can either scale or not scale. The responsibility of these Attention classes is to compute attention scores in a certain way. The scaling is a minor detail that is well defined within the responsibility of the dot attention class and reduces code complexity.
The question is how we want to expose this to the user. Right now there is "dot" and "dot-scaled" which was added quickly by me without thinking about it very long.  Maybe it might make sense to remove "dot-scaled" and allow configuring the "dot" attention to scale or not via the command line (for example by adding another flag `--rnn-scale-dot-attention` or something)?
Another way could be to annotate the DotAttention class with two registered names? Is it possible to have a 2:1 mapping with this new type of instantiation?
@tdomhan what's your opinion?

Reply:
conceptually having a single DotAttention where the scale can optionally be given would make the most sense. The only downside is that this would actually require a major version bump. If we agree that this the way to go, this should be coordinated with the top-k lexicon PR from Felix to just have a single bump for both features.

Reply:
Let me try once again to argue against single class. As everyone agrees, we have two types of attention, normal and scaled. Scaled requires a parameter, which is by default calculated as `num_hidden ** -0.5`. Now, it is not possible to supply `scale` param via cli arguments. Attention type acts as a flag to substitute this param into `DotAttention` constructor.
So here is a contradiction to the quote:
> The responsibility of these Attention classes is to compute attention scores in a certain way. The scaling is a minor detail that is well defined within the responsibility of the dot attention class and reduces code complexity.

`DotAttention` is not responsible for the all required logic - the factory must pre-compute `scale` param, set it to `non None` only then `DotAttention` will scale. Just having a type is not enough. The complexity grew up. It is spread across two parts of code - class implementation and factory method.
There are dozens of design patterns to solve that: Decorator, Template class/method, Builder etc. Having more arguments in a constructor than less is a wrong direction. Having branching logic based on optional parameters does not reduce the complexity. 
But back to the proposed solutions:
1. **Unfortunately**, two annotations on a single class will work :). But then default scale param logic should be moved back to the factory method.
2. In case you decide to deprecate one of types, the place for `scale` parameter calculation still remains questionable.

If you all agree with drawback for the single class approach - I will go for 1st option to preserve compatibility.

Reply:
additionally `is_scaled` flag can be added to `DotAttention` instead of `scale` argument.

Reply:
I think the latter approach with the boolean CLI parameter, which I think Felix also argued for with enabling scaling through `--rnn-scale-dot-attention` is the way to go as it's the simplest solution that covers both use cases covered in the previous code versions: If it's not given we don't scale and if it is given we scale be the number of hidden units. Having the logic of how to scale in the constructor of the DotAttention class makes sense to me. It's simple, clean and works :)